src/lib/firebase.js

```diff
@@ -18,17 +19,43 @@ var firebaseConfig = {
 let fb = firebase.initializeApp(firebaseConfig);
 let db = firebase.firestore();
 
-export function updatePurchaseDate(collectionName, itemId = {}) {
+export function updatePurchaseDate(collectionName, itemId) {
   const itemRef = db.collection(collectionName).doc(itemId);
-  itemRef.update({
-    purchaseDates: firebase.firestore.FieldValue.arrayUnion(
-      firebase.firestore.Timestamp.now().toMillis(),
-    ),
-  });
+  return itemRef
+    .update({
+      purchaseDates: firebase.firestore.FieldValue.arrayUnion(
+        firebase.firestore.Timestamp.now().toMillis(),
+      ),
+    })
+    .then(function() {
+      updateFrequency(collectionName, itemId);
+    });
+}
+
+export function updateFrequency(collectionName, itemId) {
+  db.collection(collectionName)
+    .where(firebase.firestore.FieldPath.documentId(), '==', itemId)
+    .onSnapshot(function(querySnapshot) {
+      querySnapshot.docChanges().forEach(change => {
+        let purchaseDates = change.doc.data().purchaseDates;
+        if (change.type === 'added' && purchaseDates.length > 2) {
+          let calculatedPurchaseFrequency = calculateFrequency(purchaseDates);
+          updateFirestore(collectionName, itemId, {
+            frequency: calculatedPurchaseFrequency,
+          });
```

Comment:
I don't fully understand the firebase methods used in this function, but I don't think there needs to be two separate update functions (updatePurchaseDates, updateFrequency). Since we're getting the purchaseDates as soon as we open the app in `App.js`, I think we could do the checks you're doing here in `List.js`, and then we could pass collectionName, itemId and updatedFrequency all in one update function. That way, we wouldn't have to communicate multiple times with Firebase for one transaction. Let me know if this is unclear!

Reply:
By checks, I meant we can check for purchaseDates length after pushing in the new purchase date, and then performing calculatedPurchaseFrequency - either doing that in `List.js` and just passing the new frequency as an argument, or you could pass the updatedPurchaseDates array (with the new purchase date added) as an argument and do the checks/calculatedPurchaseFrequency all in the one update function.  Sorry if this is rambly - I'm having a hard time coming up with a clear way of explaining without pointing to different files.

Reply:
We added the `onSnapshot` listener because we were expei

> > Oh yeah! So because 10 is out of the bounds of 1 standard deviation, so we don't include it when calculating the frequency. It's an outlier and we don't want it to skew our frequency too far in the wrong direction. That way we don't alert you at the wrong time to buy the item again!
> 
> But isn't that number a little outside of an expected range? It should be an integer representing # of days, right?

It's definitely way out of range-- the rest of the numbers are in milliseconds while we just put 10 in as the last value. We could definitely illustrate it differently by replacing the last purchase date with a value in milliseconds that is closer in time and greater than of 1 standard deviation of the mean of that array.

Reply:
I think I understand. I think this would work, I would love to hear what @katmdev and everyone else thinks!

Reply:
Hi Lauren! I'll try to address all your concerns, but let me know if I'm misinterpreting anything or have left anything out!

We kept updatePurchaseDates and updateFrequency separate so we can reuse updateFrequency when when a user removes purchase dates added in error (I think we added this to issue 13 as per Quinn's recommendation in last week's PR). As I wasn't sure if the purchase details would be in List.js or if we would create a separate component, I wanted to keep the function as modular as possible. We will likely have to add a condition that looks for removals from the purchaseDates array but I feel like that will be nicer than creating a new function.

As for the the firebase methods, we were finding it difficult to update with current information, as running the function as part of updatePurchaseDates meant we'd receive outdated data. In order to execute this function only when the purchaseDates array for the targeted item received new information, we had to leverage the updatePurchaseDates Promise to ensure the asynchronous handling of updateFrequency function. So we're getting a snapshot of the changes made in firebase to the itemId only on updates triggered by updatePurchaseDates, confirming that something has been added to the array, and then running updateFrequency. As there was a slight delay in the hand off of information between updatePurchaseDates, firebase, App state, and List props, this seemed like the most efficient way to grab the changes and update. Because we chose to use the Timestamp firebase method, we would need to process two separate writes regardless. We thought this would be a good way to meet the AC of tying the updated frequency to the update in purchase dates without having the function perpetually listening for the change.

> When a purchase is recorded, the estimated number of days until the next purchase date should be calculated and recorded in the database

Reply:
Hi Kat!
Ah ok - makes sense to keep them separate then.  I guess what I'm wondering is why it needs to get new data back before proceeding with the second function.  Maybe I'm misunderstanding, but I wonder if we can pass through the updatedPurchaseDates as an argument to updateFrequency without making it do another communication with the database.  I'm not sure if I'm getting the firebase methods right, but something like...

```
export function updatePurchaseDate(collectionName, itemId, purchaseDates) {   //include current purchaseDates as parameter
  const itemRef = db.collection(collectionName).doc(itemId);
  const currentPurchaseDate = firebase.firestore.Timestamp.now().toMillis()   // set the new purchase date
  const newPurchaseDates = purchaseDates.append(currentPurchaseDate)  // set new purchase dates array for calculate function
  return itemRef
    .update({
      purchaseDates: firebase.firestore.FieldValue.arrayUnion(
        currentPurchaseDate,
      ),
    })
    updateFrequency(collectionname, itemId, newPurchaseDates)   // send new purchaseDates
}

export function updateFrequency(collectionName, itemId, purchaseDates) {
     let calculatedPurchaseFrequency = calculateFrequency(purchaseDates);
     updateFirestore(collectionName, itemId, {
            frequency: calculatedPurchaseFrequency,
      });
}
```
This way, there would be no asynchronous methods depending on each other.

Reply:
If we do it synchronously we run the risk of getting outdated data. We have to wait for confirmation from the database that the timestamp has been created because if we run it immediately and without grabbing the snapshot of changes, we're working with the purchaseDates field in the result from props associated with itemId which doesn't yet contain the new purchase date because of the slight lag between the firebase and the shapshot in app. And I think the firebase Timestamp is generated by their server so we'd have to wait for the return in app regardless. We tried placing this function in different places without the .then but it needs that confirmation the previous action has been completed in order to get the current timestamp.

Reply:
I tested ```const currentPurchaseDate = firebase.firestore.Timestamp.now().toMillis()``` and console logging it, and it returns the milliseconds without having to communicate with Firebase.  I think the Timestamp function is available in `Firebase.js` because we're importing `firebase/firestore`.  It might be easier to talk through together rather than through text, but right now it seems like a new date/timestamp is being created in the first function, and then it's waiting and retrieving the same data in the second function, whereas I think you could just create it ahead of time as a variable and send it to both.

Reply:
I created a test branch called `lauren-test-update` to show what I mean - it took some tweaking in `firebase.js` and `list.js` to make it work, but I think I saw it updating the purchase Dates and frequency in Firebase correctly

Reply:
Awesome! I must have misunderstood the documentation! I'll take a look right now 👍 Thank you for making the branch.

Reply:
I'm not offended if you guys want to stick with the other way, but I figured it was easier to just show some code rather than try to explain it this way, haha.

Reply:
Ok just tried it and it works! Amazing addition. This is so much cleaner!! We can even just update the function to pass both the updated frequency and the updated purchase dates array in one go!!

Reply:
Wanna take a look at `kd-ql-next-extimated-purchase-date-revised`? I added the suggested changes and I'm waiting for Quinn's review before applying them to this branch. Also I think I just completely missed that we could grab the timestamp from the front end so we ended up down a long and winding path based on my incorrect assumption so your fix really helped!